debian/video-dl-3.4.2/debian/video-dl.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "README" "" "October 2015" "" ""
Video download project\.
.
.P
 \fIhttps://travis\-ci\.org/danog/video\-dl\fR  \fIhttps://npmjs\.org/package/video\-dl\fR
.
.P
Read in Italian \fIhttps://github\.com/danog/video\-dl/blob/master/README\-IT\.md\fR
.
.P
View on Github \fIhttps://github\.com/danog/video\-dl/\fR
.
.P
Created by Daniil Gentili \fIhttp://daniil\.it\fR\.
.
.P
This project is licensed under the terms of the GPLv3 license\.
.
.P
The programs included in this project can be used to download videos from any generic site including the Italian Rai \fIhttp://rai\.tv\fR Television website (including rai replay and iframe videos), the italian Mediaset \fIhttp://mediaset\.it\fR website (including iframes like the ones on the Witty TV \fIhttp://wittytv\.it\fR website), and the LA7 \fIhttp://la7\.it\fR website\. And thanks to youtube\-dl now they support lots of other websites!
.
.P
This project features: * A Bash script that can be installed on
.
.IP "\(bu" 4
\fIAny Linux/Unix system\fR
.
.IP "\(bu" 4
\fIAndroid\fR
.
.IP "\(bu" 4
\fIMac OS X\fR
.
.IP "\(bu" 4
\fIiOS\fR and even on
.
.IP "\(bu" 4
\fIWindows\fR!
.
.IP "\(bu" 4
An \fIAPI\fR!
.
.IP "\(bu" 4
An \fIAndroid app\fR!
.
.IP "\(bu" 4
And even a \fIweb version\fR!
.
.IP "\(bu" 4
That can be \fIincorporated\fR in other websites!
.
.IP "" 0
.
.P
Both the \fIAPI\fR and the \fIweb version\fR use a database \fIhttps://github\.com/danog/video\-dl/blob/master/video_db\.sql\fR\.This project also features a web version \fIhttps://video\.daniil\.it/\fR\.
.
.P
.
.P
The source code of the page can be viewed here \fIhttps://github\.com/danog/video\-dl/blob/master/web\fR\.
.
.P
I used the following programs in the web version:
.
.IP "\(bu" 4
Loading animation (pace\.js) \fIhttps://github\.com/HubSpot/pace\fR
.
.IP "\(bu" 4
HTML entity encoder (he\.js) \fIhttps://github\.com/mathiasbynens/he\fR
.
.IP "\(bu" 4
Linkifier (linkify\.js) \fIhttps://github\.com/soapbox/linkifyjs/\fR
.
.IP "" 0
.
.P
And used the following theme as a base:
.
.IP "\(bu" 4
Bootstrap theme (freelancer) \fIhttps://github\.com/IronSummitMedia/startbootstrap\-freelancer\fR
.
.IP "" 0
.
.SH "Incorporation"
You can incorporate this script in your website!
.
.br
Just include the video\-dl script with:
.
.br
```
.
.IP "" 4
.
.nf

Or install it using npm\.
.
.fi
.
.IP "" 0
.
.P
npm install video\-dl ```
.
.P
Here\'s a list of the functions and the usage instructions\.
.
.SS "Video Download function\."
.
.IP "" 4
.
.nf

video_dl(output, inputurl, dlsupport, messageoutput)
.
.fi
.
.IP "" 0
.
.P
example: \fB#result\.\fR
.
.P
example: \fB$("input#urljs")\.val()\fR
.
.P
0 enables, anything disables\.
.
.P
example: \fB#message\fR
.
.P
\fBvideo_dl("#result", $("input#urljs")\.val(), "0", "#message");\fR
.
.P
Let\'s say the \fBinput#urljs\fR text field has value \fB"http://www\.winx\.rai\.it/dl/RaiTV/programmi/media/ContentItem\-a27ccfe8\-b824\-4e85\-9a08\-d15e57fb61a0\.html#p=0"\fR\.
.
.P
The function will get the value of the \fBinput#urljs\fR element, get the download links from the API, and return the following output to the \fB#result\fR element:
.
.br
```
.
.br
.
.br
.
.br
.
.br
```
.
.P
The function will then start the mailtext function with the following parameters:
.
.br
\fBmailtext(messageoutput, inputurl);\fR
.
.P
See mailtext description for the result\.
.
.SS "First load function\."
Use to output supported websites list, prepare mail message and, only on the video\.daniil\.it website, hide the php submit module and use javascript engine instead\.
.
.IP "" 4
.
.nf

firstload(supportedurls, separatorstart, separatorend, messageoutput, videodaniilit)
.
.fi
.
.IP "" 0
.
.P
Example: \fB#supportedurls\fR\.
.
.P
Optional, recommended\.
.
.br
Example: \fB<li>\fR
.
.P
Example: \fB</li>\fR
.
.P
Example: \fB#contact\fR
.
.IP "" 4
.
.nf

firstload("#supportedurls", "<li>", "</li>", "#message");
.
.fi
.
.IP "" 0
.
.P
Let\'s say the url list is: \fBa b c d\fR\.
.
.P
Output printed to #supportedurls is:
.
.IP "" 4
.
.nf

<li>a</li><li>b</li><li>c</li><li>d</li><a href="http://lol\.daniil\.it" target="_blank">&#9786;</a></li>
.
.fi
.
.IP "" 0
.
.P
This will also create the default contact module text with \fBmailtext("#message");\fR
.
.SS "Contact module function\."
Prints a nice message to the contact module text field, with the url if it\'s provided else just With \fBinsert link\fR\.
.
.IP "" 4
.
.nf

mailtext(output, url)
.
.fi
.
.IP "" 0
.
.P
Example: \fB#contact\fR
.
.IP "" 4
.
.nf

mailtext("#contact", "http://google\.com");
.
.fi
.
.IP "" 0
.
.P
Will put \fBThe video: http://google\.com does not download, could you please fix it Thanks!\fR
.
.P
to \fB#contact\fR\.
.
.SS "Error function\."
Sends me an email if an error occurs\.
.
.IP "" 4
.
.nf

error(output, url, error)
.
.fi
.
.IP "" 0
.
.P
example: \fB#result\fR
.
.P
example: \fBhttp://google\.com\fR
.
.P
Empty Response\. Recommended\.
.
.br
example: \fBerror 404\fR
.
.IP "" 4
.
.nf

error("#result", "http://google\.com", "error 404");
.
.fi
.
.IP "" 0
.
.P
Will try to send me a mail and print \fB"An error occurred and it was reported!"\fR to \fB#result```` if everything went fine, else it will print\fR"An error occurred but it couldn\'t be reported! Please use the manual report module!"```\.
.
.SH "Bash script usage:"
.
.nf

video\.sh [ \-qaf [ urls\.txt ] ] URL URL2 URL3 \.\.\.

Do not forget to put the URL between quotes if it contains special chars like & or #\.

Run with \./video\.sh if you installed in a directory not in $PATH\.

Options:




\-q              Quiet mode: useful for crontab jobs, automatically enables \-a\.

\-a              Automatic mode: automatically download the video in the maximum quality\.

\-b              Use built\-in API engine: requires additional programs and may not work properly on some systems but may be faster than the API server\.

\-f              Read URL(s) from specified text file(s)\. If specified, you cannot provide URLs as arguments\.

\-p=player       Play the video instead of downloading it using specified player, mplayer if none specified\.

\-\-help          Show this extremely helpful message\.
.
.fi
.
.SS "Bash script installation instructions:"
.
.P
On debian\-derived distros, execute this command to add my repo to your system:
.
.IP "" 4
.
.nf

sudo wget \-q \-O /etc/apt/sources\.list\.d/daniil\.list http://dano\.cu\.cc/1IJrcd1 && wget \-q \-O \- http://dano\.cu\.cc/1Aci9Qp | sudo apt\-key add \- && sudo apt\-key adv \-\-recv\-keys \-\-keyserver keys\.gnupg\.net 72B97FD1D9672C93 && sudo apt\-get update
.
.fi
.
.IP "" 0
.
.P
You should see an OK if the operation was successful\.
.
.P
And this command to install the script\.
.
.IP "" 4
.
.nf

sudo apt\-get update; sudo apt\-get \-y install video\-dl
.
.fi
.
.IP "" 0
.
.P
Execute this command to install the script:
.
.IP "" 4
.
.nf

wget http://daniilgentili\.magix\.net/video\.sh \-O video\.sh || curl \-L http://daniilgentili\.magix\.net/video\.sh \-o video\.sh; chmod +x video\.sh
.
.fi
.
.IP "" 0
.
.P
Run with \./video\.sh in the directory where you downloaded it\.
.
.P
To use from any directory install the script directly in the $PATH using this command (run as root):
.
.IP "" 4
.
.nf

wget http://daniilgentili\.magix\.net/video\.sh \-O /usr/bin/video\.sh || curl \-L http://daniilgentili\.magix\.net/video\.sh \-o video\.sh; chmod +x /usr/bin/video\.sh
.
.fi
.
.IP "" 0
.
.P
Enable unknown sources and install this app \fIhttp://bit\.ly/0192837465k\fR\. Once opened you will be presented with a user friendly interface similar to the web version\.
.
.P
1: initial version
.
.P
1\.2: added not working, share and credits button
.
.P
1\.2\.1: added external sharing option, fixed bugs
.
.P
1\.2\.2: Fixed not working button on external share URL, added google analytics, fixed Rai Replay on external share\.
.
.P
1\.3: Added auto update\.
.
.P
1\.4: Added more credits\.
.
.P
You tell me!
.
.P
Video tutorial \fIhttps://www\.youtube\.com/watch?v=4NLs2NzHbbc\fR
.
.P
Note: if you can\'t copy & paste the commands directly in the Terminal Emulator app try this: paste them in the url bar one line at a time, copy them again from the url bar and try to paste them again in the Terminal Emulator app\. Run these commands: \fBcd /sdcard && wget http://daniilgentili\.magix\.net/android/video\.sh\fR
.
.P
Run with: \fBbash /sdcard/video\.sh\fR
.
.P
To install the script directly in the $PATH use these commands (here, root is mandatory)\.
.
.IP "" 4
.
.nf

su
mount \-o rw,remount /system && wget http://daniilgentili\.magix\.net/android/video\.sh \-O /system/bin/video\.sh && chmod 755 /system/bin/video\.sh
.
.fi
.
.IP "" 0
.
.P
If you cannot execute the script match the its shebang (the #!) to the location of the bash executable\.
.
.SS "iOS:"
Jailbreak your device, add the following repo to Cydia,
.
.IP "" 4
.
.nf

http://repo\.daniil\.it
.
.fi
.
.IP "" 0
.
.P
\&\.\.\. and install mobileterminal and video\-dl\.
.
.P
Run with video\.sh in mobileterminal\.
.
.P
To view and import the downloaded video to the gallery use iFile or Filza\.
.
.SS "Windows:"
Install Cygwin \fIhttps://www\.cygwin\.com\fR (don\'t forget to install wget and sed during the installation process!), open its command prompt and type:
.
.IP "" 4
.
.nf

wget http://daniilgentili\.magix\.net/win/video\.sh \-O video\.sh
.
.fi
.
.IP "" 0
.
.P
Run with \./video\.sh In the directory where you downloaded it\.
.
.P
To run the script from any directory run the following commands:
.
.IP "" 4
.
.nf

cd /bin && wget http://daniilgentili\.magix\.net/win/video\.sh \-O video\.sh && cd $OLDPWD
.
.fi
.
.IP "" 0
.
.SH "API"
This project also features an API \fIhttp://video\.daniil\.it/api/\fR\.
.
.P
The source code of the API can be viewed here \fIhttps://github\.com/danog/video\-dl/blob/master/api\fR\.
.
.P
The API uses youtube\-dl \fIhttps://github\.com/rg3/youtube\-dl\fR to get the links for non Rai/mediaset/la7 videos\.
.
.P
The API supports GET requests and the endpoint is http://api\.daniil\.it (supports https)\.
.
.SS "Supported parameters:"
.
.P
The value should be the URL of the video to download\. The response will be a list of URLS with the corresponding quality name, format, size and dimension\.
.
.P
Example:
.
.IP "" 4
.
.nf

http://api\.daniil\.it/?url=http://www\.winx\.rai\.it/dl/RaiTV/programmi/media/ContentItem\-47307196\-8fd1\-46f8\-8b31\-92ae5f9b5089\.html#p=0
.
.fi
.
.IP "" 0
.
.P
Output:
.
.IP "" 4
.
.nf

Winx_Club_VI_Ep3_Il_collegio_volante Winx Club VI \- Ep\.3: Il collegio volante
Highest quality (mp4, 286MB, 1024x576) http://creativemedia4\.rai\.it/Italy/podcastcdn/junior/Winx/Winx_6_puntate/2189463_1800\.mp4
Medium\-low quality (mp4, 131MB, 700x394) http://creativemedia4\.rai\.it/Italy/podcastcdn/junior/Winx/Winx_6_puntate/2189463_800\.mp4
.
.fi
.
.IP "" 0
.
.P
Explanation:
.
.IP "" 4
.
.nf

Winx_Club_VI_Ep3_Il_collegio_volante Winx Club VI \- Ep\.3: Il collegio volante
.
.fi
.
.IP "" 0
.
.P
Sanitized name of video for file name Original name of the video for printing to user output
.
.IP "" 4
.
.nf

Highest quality (mp4, 286MB, 1024x576) http://creativemedia4\.rai\.it/Italy/podcastcdn/junior/Winx/Winx_6_puntate/2189463_1800\.mp4
.
.fi
.
.IP "" 0
.
.P
Quality name (format, size, dimension) URL of the video
.
.IP "" 4
.
.nf

Medium\-low quality (mp4, 131MB, 700x394) http://creativemedia4\.rai\.it/Italy/podcastcdn/junior/Winx/Winx_6_puntate/2189463_800\.mp4
.
.fi
.
.IP "" 0
.
.P
Quality name (format, size, dimension) URL of the video
.
.P
The qualities are ordered in decreasing order by dimension\.
.
.P
Supports the following values:
.
.IP "" 4
.
.nf

websites
.
.fi
.
.IP "" 0
.
.P
returns a shortened list of supported websites:
.
.IP "" 4
.
.nf

allwebsites
.
.fi
.
.IP "" 0
.
.P
returns a full list of supported websites\.
.
.SH "Contribute!"
If you created a version of the script using another programming language contact me \fIhttp://daniil\.it\fR and I will put it on this page!
.
.P
That\'s it!
.
.P
Enjoy!
.
.P
Daniil Gentili \fIhttp://daniil\.it\fR


video-dl.js
/* Video download script, jquery version - Copyright (C) 2015 Daniil Gentili
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it under certain conditions; see https://github.com/danog/video-dl/raw/master/LICENSE.

Video Download function

Usage:
video_dl(output, inputurl, dlsupport, messageoutput)

Parameters:

output: output html element for the video info and the download urls. Required.
example: #result


inputurl: url of the video. Required.
example: $("input#urljs").val()

dlsupport: enables or disables the download attribute in the download links. Optional.
0 enables, anything disables

messageoutput: output html element for the contact module. Optional.
example: #message

Example: 
video_dl("#result", $("input#urljs").val(), "0", "#message");

Let's say the input#urljs text field has value "http://www.winx.rai.it/dl/RaiTV/programmi/media/ContentItem-a27ccfe8-b824-4e85-9a08-d15e57fb61a0.html#p=0".

The function will get the value of the input#urljs element, get the download links from the API, and return the following output to the #result element:

<h1 style="font-style: italic;">Video download script.</h1><br><h2 style="font-style: italic;">Created by <a href="http://daniil.it">Daniil Gentili</a></h2><br><h1>Title:</h1> <h2>26 - Il potere degli animali fatati - Winx Club VII del 03/10/2015</h2><br><h1>Available versions:</h1><br><h2><a href="http://creativemedia4.rai.it/Italy/podcastcdn/junior/Winx/Winx_7_EP_Puntate/4524680.mp4" download="26_Il_potere_degli_animali_fatati_Winx_Club_VII_del_03102015.mp4">Normal quality (mp4, 267 MiB, 720x404)</a><br></h2>

The function will then start the mailtext function with the following parameters:
mailtext(messageoutput, inputurl);

See mailtext description for the result.
*/

function video_dl(output, inputurl, dlsupport, messageoutput) {
    // Console log for debugging
    console.log("input is "+inputurl+", output is "+output+" and messageoutput is "+messageoutput);
    // Check if output param exists. 
    if (!(output)) { console.log("Missing output parameter.");return; };
    // Clear output html element
    $(output).empty();
    // Check if URL is provided
    if (inputurl) {
        // Working...
        $(output).html("<h2>Working...</h2>");
        // Prepare URL for API request
        url = "https://api.daniil.it/?url=" + encodeURIComponent(inputurl);
        // Create XMLHttpRequest 
        var xmlhttp = new XMLHttpRequest();
        // Prepare request
        xmlhttp.onreadystatechange = function() {
            // When done,
            if (xmlhttp.readyState == 4 || xmlhttp.readyState == "complete") {
                // Get the response
                response = xmlhttp.responseText;
                // And prepare the output
                if (response) {
                    // Encode response in html characters, just in case
                    response = he.encode(response);
                    // Get the titles
                    var titles = response.substr(0, response.indexOf("\n"));
                    // Get the sanitized title
                    var title = titles.substr(0, response.indexOf(" "));
                    // Get the non sanitized title
                    var videoTitolo = titles.substring(response.indexOf(" ") + 1);
                    // Prepare first part of output
                    var result = "<h1><i>Video download script.</i></h1><br><h2><i>Created by <a href=\"http://daniil.it\">Daniil Gentili</a></i></h2><br><h1>Title:</h1> <h2>" +videoTitolo+ "</h2><br><h1>Available versions:</h1>";
                    // Remove the titles, keep urls
                    var lines = response.split('\n');
                    lines.splice(0, 1);
                    // Loop trough the lines (qualities)
                    for (var i = 0; i < lines.length; i++) {
                        // Get info
                        last = lines[i].lastIndexOf(" ");
                        info = lines[i].substring(0, last);
                        // Get url
                        splitr = lines[i].split(" ");
                        videourl = splitr[splitr.length - 1];
                        // Choose wherether or not to use the download attribute
                        if(dlsupport === "0") {
                            // Get video extension from info
                            ext = info.substring(info.indexOf('(') + 1);
                            ext = ext.substring(0, ext.indexOf(','));
                            // Create dl variable
                            dl = " download=\"" + title + "." + ext + "\"" ;
                            // Or dont.
                        } else { var dl = ""; };
                        // Append result to final result
                        result += "<h2><a"+ dl + " href=\"" + videourl + "\">" + info + "</a></h2><br>"
                    }
                    // Output the result
                    $(output).html(result);
                    // Else send error email
                } else error(output, inputurl, response);
            }
        }
        // Set request type
        xmlhttp.open("GET", url, true);
        // Send request
        xmlhttp.send();
        // Put url to contact module
        if (messageoutput) { mailtext(messageoutput, inputurl); };
        // If no URL was provided
    } else { $(output).html("<h1>No URL was provided!</h1>"); }
};

/*
First load function, used to output supported websites list, prepare mail message and,
only on the video.daniil.it website, hide the php submit module and use javascript engine instead.

Usage: firstload(supportedurls, separatorstart, separatorend, messageoutput, videodaniilit)

Parameters:

supportedurls: output html element for the supported websites list. Required.
Example: #supportedurls

separatorstart: first separator for the supported urls list: it will be put before every item in the supported websites list, if empty defaults to <br>.
Optional, recommended.
Example: <li>

separatorend: the second separator for the supported urls list: it will be put after every url, if empty defaults to <br>. Optional, recommended.
Example: </li>

messageoutput: output html element for default contact module text. Optional, recommended.
Example: #contact

videodaniilit: If on video.daniiil.it hides php module and unhides javascript text field. Do not use.

Example:
firstload("#supportedurls", "<li>", "</li>", "#message");

Let's say the url list is: a b c d

Output printed to #supportedurls is:
<li>a</li><li>b</li><li>c</li><li>d</li><a href=\"http://lol.daniil.it\" target=\"_blank\">&#9786;</a></li>
This will also create the default contact module text with
mailtext("#message"); 

*/

function firstload(supportedurls, separatorstart, separatorend, messageoutput, videodaniilit) {

    // If no separator is set, set them to <BR>
    if (!((separatorstart) && (separatorend))) { var separatorstart = "<br>"; var separatorstart = separatorend; };

    // Clear supportedurls element
    $(supportedurls).empty();
    // Write working... to supportedurls element
    $(supportedurls).html("<h3>Working...</h3>");

    // Set request url
    url = "https://api.daniil.it/?p=allwebsites";

    // Create request
    var xmlhttp = new XMLHttpRequest();

    // What to do after request success
    xmlhttp.onreadystatechange = function() {
        if (xmlhttp.readyState == 4 || xmlhttp.readyState == "complete") {
            // Set response text
            response = xmlhttp.responseText;
            // If response isn't empty
            if (response) {
                // If on video.daniiil.it hide php module and unhide javascript text field.

                if (videodaniilit) {
                    $("#js").css("display", "block");
                    $("#php").css("display", "none");
                    $("#jsd").css("display", "block");
                    $("#phpd").css("display", "none");
                };
                // Replace newlines in response with separators
                response = he.encode(response).replace(/\n/g, separatorend+separatorstart);

                // Output the result and the mail text
                $(supportedurls).html(separatorstart+response+"<a href=\"http://lol.daniil.it\" target=\"_blank\">&#9786;</a>"+separatorend);
                // Linkify output
                $(supportedurls).linkify({
                    target: "_blank"
                });
            }
        }
    }
    // Setup request
    xmlhttp.open("GET", url, true);
    // Send request
    xmlhttp.send();
    // Create default contact module text
    mailtext(messageoutput);

};

/*
Contact module function.
Prints a nice message to the contact module text field, with the url if it's provided else just With "insert link".

Usage: 
mailtext(output, url)

Parameters:

output: html entity where to print out the contact message. Required.
Example: #contact

url: url of the video to insert into the message. Not required, if not provided defaults to insert link.

Example:
mailtext("#contact", "http://google.com");

Will put "The video:
http://google.com
does not download, could you please fix it
Thanks!" to #contact.
*/
function mailtext(output, url) {
    // Check for required params.
    if (!(output)) { console.log("Missing required params.");return; };

    // If url isn't empty, put it in the message
    if (url) {
        mail = url;
    } else {
        // Else put insert link instead.
        mail = "insert link";
    }

    // Create mail message
    var mailmessage = 'The video:\n' + mail + '\ndoes not download, could you please fix it?\nThanks!';

    // Insert mail message in page
    $(output).html(mailmessage);
};


/*
Error function.
Sends me an email if an error occurs.

Usage:
error(output, url, error)

Parameters:

output: html entity where to output success or error message of request to mail sending php script. Required.
example: #result

url: url of the video that failed to download. Required.
example: http://google.com

error: actual error message. If empty defaults to
Empty Response. Recommended.
example: error 404

Example:
error("#result", "http://google.com", "error 404");

Will try to send me a mail and print "An error occurred and it was reported!" to #result if everything went fine, else it will print "An error occurred but it couldn't be reported! Please use the manual report module!".

*/
function error(output, url, error) {
    // Check for required params
    if (!((url) && (output))) { console.log("Missing required params.");return; };
    // Set domain value for php script
    var domain = 3;
    // Log to console the error.
    console.log(error);
    // If there is no error set it to Empty Response
    if (!(error)) { var error = "Empty Response"; };
    // Prepare and make request
    $.ajax({
        url: "https://mail.daniil.it/",
        type: "POST",
        data: {
            url: url,
            error: error,
            domain: domain
        },
        cache: false,
        success: function() {
            // Success sending mail
            $(output).html("<h1>An error occurred and it was reported!</h1>");
        },
        error: function() {
            // Error sending mail
            $(output).html("<h1>An error occurred but it couldn't be reported! Please use the manual report module!</h1>");
        },
    })
};


